pkgs/stdenv/generic/check-meta.nix

```diff
@@ -453,7 +453,32 @@ let
       outputs = attrs.outputs or [ "out" ];
       hasOutput = out: builtins.elem out outputs;
     in
-    {
+      # only try to set the default if relevant attributes are present.
+      # additionaly, several packages assume src will only be evaluated on
+      # supported platforms, so skip if the package is unsupported
+      optionalAttrs (!validity.unsupported && attrs ? src.meta.homepage || attrs ? srcs && isList attrs.srcs && any (src: src ? meta.homepage) attrs.srcs) {
+      # should point to an http-browsable source tree, if available.
+      # fetchers like fetchFromGitHub set it automatically.
+      # this could be handled a lot easier if we nulled it instead
+      # of having it be undefined, but that wouldn't match the
+      # other attributes.
+      repository = let
+        getSrcs = attrs:
+          if attrs ? src
+          then
+            [ attrs.src ]
+          else
+            lib.filter (src: src ? meta.homepage) attrs.srcs;
+        getHomePages = srcs: map (src: src.meta.homepage) srcs;
+        unlist = list:
+          if lib.length list == 1
+          then
+            lib.elemAt list 0
+          else
+            list;
+      in
+        unlist (getHomePages (getSrcs attrs));
+    } // {
```

Comment:
Since the user is somewhat free to set `meta.repository`, the use of `//` can bring unexpected outcomes.

https://nix.dev/guides/best-practices.html#updating-nested-attribute-sets

It's better to run this logic inside the block (below `name = . . .;` below),  somewhat like

```nix
repository = 
  if (complicatedConditions) then
   complicatedFunction
  else
    [ ];
```

Reply:
> Since the user is somewhat free to set meta.repository, the use of // can bring unexpected outcomes.

what unexpected outcomes?

manually setting `repository` will override the autodetected value, but that is intended behavior.

that link specifically warns of the fact that `//` performs a shallow update, but a shallow update is what is desired here.

> It's better to run this logic inside the block (below name = . . .; below), somewhat like

if you want the field to default to unset (to match the behavior of the other fields), that is impossible, since nix has no equivalent to javascript's `undefined`.

Reply:
> if you want the field to default to unset

no. This is only strange.

But the surrounding code is no less strange.

Reply:
if you have a problem with defaulting to unset, take it up with the original authors of `check-meta.nix` or open a github issue, it wasn't my choice.

Reply:
I DO NOT LIKE DEFAULT TO UNSET

Can you please stop trying to read what I did not say?

Reply:
**I never said a thing about** "default to unset".

Reply:
> > if you want the field to default to unset
> 
> no. This is only strange.
> 
> But the surrounding code is no less strange.

is this not saying "defaulting to unset is strange"?

if that's not what you are trying to say, please elaborate, i can't quite understand that sentence.

Reply:
"No" means "This is not what was I talking about".

"Strange" is, more precisely, about the fact the code looks like "I need to grab a piece of information, but that piece of information is hidden after a long and fragile matrioska of nested attributes that can break in unexpected ways".

It should not be so hard to just query `src.meta.homepage`.

On the other hand, the whole check-meta file is equally hard to understand. That's what I meant.

Reply:
most of check-meta actually has to do with reccomending remediations for how to fix common errors (such as enabling NIXPKGS_ALLOW_UNFREE)

the actual verification is relatively simple.

> It should not be so hard to just query src.meta.homepage

no, but unfortunately we are running into a significant downside of lazy evaluation here, and there's not a ton we can do.

Reply:
also a downside of not having implicit null

Reply:
eliminating the optionalAttrs also eliminates the eval performance hit. https://github.com/NixOS/nixpkgs/pull/300177#issuecomment-2029698758

Reply:
hmm... i guess we could just make it default to the empty list and be a bit different.  i believe nixos search treats unset as an empty list anyways.

Reply:
additionally, not using // means the conditions are only evaluated if meta.repository is not already set, which means any packages that have a src that doesn't eval can just set `repository` manually.